### ISSUE TITLE
Run half-precision models on Android

### ISSUE THREAD

**Issue by hicham-eyeem** [tensorflow/tensorflow#9664]
Is it possible to run half-precision (float16) graphs on Android (arm64-v8a/AArch64 supports half-precision)? If so, what would be the approach to do that? Trying to run a float16 graph gives an exception saying that only INT32 and Float32 ops are supported.

Thank you in advance,

**Comment by drpngx** [comment]
@andrewharp do you know how far we are from that?

**Comment by andrewharp** [comment]
@petewarden can correct me if I'm off base here, but I don't believe float16 support is something we're currently targeting.

Most of our attention on mobile is focused on quantized 8bit support, which should give you even better size gains and possibly performance as well. Support for quantizing graphs can be found in the [transform_graph](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md) tool.

**Comment by hicham-eyeem** [comment]
@andrewharp thank you for the feedback. I am familiar with the 8bit quantization tool. Unfortunately, it results in a major drop in performance in my case. Probably because the network uses a custom activation (not ReLU, but something similar to ELU) while the quantization tool is (I believe) optimized for the ReLU activation. Also, I would like to use a higher precision than 8-bit for my problem, hence my interest in the half-precision support.
But since Eigen supports half-precision, I believe half-precision support would be relatively easier to integrate than 8bit quantization since it does not require any fancy op mapping.

**Comment by Androbin** [comment]
From the description of the `quantize_weights` transform:
> Even though there are no prerequesites, it is advisable to run fold_batch_norms or fold_old_batch_norms, because rounding variances down to zero may cause significant loss of precision.

Could it be that your loss in precision is (similar to) this kind of numerical hiccup?

**Comment by hicham-eyeem** [comment]
@Androbin Thank you for the feedback but I don't use batch normalization. I think the main problem with the quantization tool is that the activation should be ReLU (the rounding option works fine though).

**Comment by andrewharp** [comment]
@hicham-eyeem As this is an issue involving a custom op we don't really have the bandwidth to investigate right now, but if you want to propose a general solution in a PR we'd be happy to take a look.

**Comment by concretevitamin** [comment]
(Triaging issues.) Since @andrewharp from the TF Mobile team is willing to take a look at PRs, I'm marking this as contributions welcome.

**Comment by hicham-eyeem** [comment]
@andrewharp thank you, I hope I will have time to take a look at it. I remember I have also tried to decompose ELU as a sum of ReLU and masked exp(input) - 1 , still produced the same results, so I think the problem comes from the exp function on quantized values.

**Comment by 879622409** [comment]
I got the same problem. Is there any solution?

**Comment by mohantym** [comment]
Hi @hicham-eyeem ! Does above [PR](https://github.com/tensorflow/tensorflow/pull/53274) resolve this issue?

**Comment by tilakrayal** [comment]
@hicham-eyeem,
Could you please take a look at this [PR](https://github.com/tensorflow/tensorflow/pull/53274) and confirm whether the PR resolves the issue. Thank you!
